waiter/src/waiter/service_description.clj

```diff
@@ -340,7 +362,20 @@
        :service-id service-id}))
 
   (validate [_ service-description args-map]
-    (validate-schema service-description (merge-with set/union args-map {:valid-cmd-types #{"shell"}}))))
+    (->> (merge-with set/union args-map {:valid-cmd-types #{"shell"}})
+         (validate-schema service-description upper-limits-schema))))
+
+(defn create-default-service-description-builder
+  "Returns a new DefaultServiceDescriptionBuilder which uses the specified resource limits."
+  [{:keys [constraints]}]
+  (let [upper-limits (->> constraints
+                          (filter (fn [[_ constraint]] (contains? constraint :max)))
+                          (pc/map-vals :max))
+        upper-limits-schema (-> (->> upper-limits
+                                     (pc/map-keys s/optional-key)
+                                     (pc/map-vals (fn [v] (s/pred #(<= % v) (symbol (str "limit-" v))))))
+                                (assoc s/Str s/Any))]
+    (->DefaultServiceDescriptionBuilder upper-limits-schema)))
```

Comment:
`max-constraints-schema`?

Reply:
Let's be consistent.  `max` instead of `upper`, `constraint` instead of `limit` throughout

Reply:
Done.